packages/server/lib/modes/run.ts

```diff
@@ -0,0 +1,1661 @@
+
+    if (startedVideoCapture && !videoExists) {
+      // the video file no longer exists at the path where we expect it,
+      // likely because the user deleted it in the after:spec event
+      debug(`No video found after spec ran - skipping processing. Video path: ${videoName}`)
+
+      results.video = null
+    }
+
+    const hasFailingTests = _.get(stats, 'failures') > 0
+    // we should upload the video if we upload on passes (by default)
+    // or if we have any failures and have started the video
+    const shouldUploadVideo = !skippedSpec && videoUploadOnPasses === true || Boolean((startedVideoCapture && hasFailingTests))
+
+    results.shouldUploadVideo = shouldUploadVideo
+
+    if (!quiet && !skippedSpec) {
+      displayResults(results, estimated)
+      if (screenshots && screenshots.length) {
+        displayScreenshots(screenshots)
+      }
+    }
+
+    const project = openProject.getProject()
+
+    if (!project) throw new Error('Missing project!')
+
+    // @ts-expect-error experimentalSingleTabRunMode only exists on the CT-specific config type
+    const usingExperimentalSingleTabMode = testingType === 'component' && config.experimentalSingleTabRunMode
+
+    if (usingExperimentalSingleTabMode) {
+      await project.server.destroyAut()
+    }
+
+    // we do not support experimentalSingleTabRunMode for e2e
+    if (!usingExperimentalSingleTabMode) {
+      debug('attempting to close the browser tab')
+
+      await openProject.resetBrowserTabsForNextTest(shouldKeepTabOpen)
+
+      debug('resetting server state')
+
+      project.server.reset()
+    }
+
+    if (videoExists && !skippedSpec && endVideoCapture && !videoCaptureFailed) {
+      const ffmpegChaptersConfig = videoCapture.generateFfmpegChaptersConfig(results.tests)
+
+      await postProcessRecording(
+        videoName,
+        compressedVideoName,
+        videoCompression,
+        shouldUploadVideo,
+        quiet,
+        ffmpegChaptersConfig,
+      )
+      .catch(warnVideoRecordingFailed)
+    }
+
+    return results
+  })
+}
+
+function screenshotMetadata (data, resp) {
+  return {
+    screenshotId: random.id(),
+    name: data.name || null,
+    testId: data.testId,
+    testAttemptIndex: data.testAttemptIndex,
+    takenAt: resp.takenAt,
+    path: resp.path,
+    height: resp.dimensions.height,
+    width: resp.dimensions.width,
+  }
+}
+
+async function runSpecs (options: { config: Cfg, browser: Browser, sys: any, headed: boolean, outputPath: string, specs: SpecWithRelativeRoot[], specPattern: string | RegExp | string[], beforeSpecRun?: BeforeSpecRun, afterSpecRun?: AfterSpecRun, runUrl?: string, parallel?: boolean, group?: string, tag?: string, testingType: TestingType, quiet: boolean, project: Project, onError: (err: Error) => void, exit: boolean, socketId: string, webSecurity: boolean, projectRoot: string } & Pick<Cfg, 'video' | 'videoCompression' | 'videosFolder' | 'videoUploadOnPasses'>) {
```

Comment:
Ahhh yes awesome, we need this so desperately. Every time I tried to make this into TS, I realized how much work it was and abandoned it - definitely time well spent to figure out what these types are.

Reply:
Yeah, there's still about 4 billion de-structured variables too many in this file, but at least this will enable us to more confidently refactor in the future.